### PR TITLE
[RyuJIT] Fold "(X op C1) op C2" to "X op (C1 op C2)" for commutative operators

### LINKED ISSUE #43551: JIT fails to constant fold multiplies, and/or merge shifts
```C#
    [MethodImpl(MethodImplOptions.NoInlining)]
    static int Y(int x) => x * 2 * 2;
```
produces
```asm
; Assembly listing for method X:Y(int):int
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  3,  3   )     int  ->  rcx
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]   "OutgoingArgSpace"
;
; Lcl frame size = 0

G_M5754_IG01:              ;; offset=0000H
                                                ;; bbWeight=1    PerfScore 0.00
G_M5754_IG02:              ;; offset=0000H
       8D0409               lea      eax, [rcx+rcx]
       03C0                 add      eax, eax
                                                ;; bbWeight=1    PerfScore 0.75
G_M5754_IG03:              ;; offset=0005H
       C3                   ret
                                                ;; bbWeight=1    PerfScore 1.00

; Total bytes of code 6, prolog size 0, PerfScore 2.35, instruction count 3 (MethodHash=7c40e985) for method X:Y(int):int
```

Should be simple enough to handle this in morph. Or we could re-associate better.

category:cq
theme:expression-opts
skill-level:intermediate
cost:medium

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/43551

Folds `(X op C1) op C2` for `+`, `*`, `&`, `|`, `^` operators to `X op (C1 op C2)` where `C1 op C2` is folded.
Most popular case for this optimization is:

```csharp
MyFlags Test(MyFlags x) => x | MyFlags.Flag1 | MyFlags.Flag2;
```
Current codegen:
```asm
       8BC2                 mov      eax, edx
       83C808               or       eax, 8
       83C840               or       eax, 64
       C3                   ret 
```
New codegen:
```asm
       8BC2                 mov      eax, edx
       83C848               or       eax, 72
       C3                   ret 
```

Jit-diff:
```
C:\prj>jit-diff diff --output C:\prj\jitdiffs -f --core_root C:\prj\runtime-1\artifacts\tests\coreclr\Windows_NT.x64.Release\Tests\Core_Root --base C:\prj\runtime-1\artifacts\bin\coreclr\Windows_NT.x64.Checked_base --diff C:\prj\runtime-1\artifacts\bin\coreclr\Windows_NT.x64.Checked --crossgen C:\prj\runtime-1\artifacts\tests\coreclr\Windows_NT.x64.Release\Tests\Core_Root\crossgen.exe
Beginning Crossgen CodeSize Diffs for System.Private.CoreLib.dll, framework assemblies
| Finished 268/268 Base 268/268 Diff [113.6 sec]
Completed Crossgen CodeSize Diffs for System.Private.CoreLib.dll, framework assemblies in 113.72s
Diffs (if any) can be viewed by comparing: C:\prj\jitdiffs\dasmset_19\base C:\prj\jitdiffs\dasmset_19\diff
Analyzing CodeSize diffs...
Found 13 files with textual diffs.
Crossgen CodeSize Diffs for System.Private.CoreLib.dll, framework assemblies for  default jit
Summary of Code Size diffs:
(Lower is better)
Total bytes of diff: -815 (-0.00% of base)
    diff is an improvement.
Top file improvements (bytes):
        -423 : Microsoft.CodeAnalysis.CSharp.dasm (-0.02% of base)
        -184 : System.Private.CoreLib.dasm (-0.01% of base)
         -73 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-0.00% of base)
         -42 : System.DirectoryServices.dasm (-0.01% of base)
         -23 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.00% of base)
         -19 : System.Private.Uri.dasm (-0.02% of base)
         -13 : System.Private.Xml.dasm (-0.00% of base)
         -12 : System.DirectoryServices.AccountManagement.dasm (-0.00% of base)
         -10 : System.Memory.dasm (-0.01% of base)
          -6 : System.Runtime.Serialization.Formatters.dasm (-0.01% of base)
          -4 : Microsoft.VisualBasic.Core.dasm (-0.00% of base)
          -3 : System.Data.Common.dasm (-0.00% of base)
          -3 : System.Net.Security.dasm (-0.00% of base)
13 total files with Code Size differences (13 improved, 0 regressed), 255 unchanged.
Top method regressions (bytes):
           3 ( 0.06% of base) : System.Private.CoreLib.dasm - RuntimeType:InvokeMember(String,int,Binder,Object,ref,ref,CultureInfo,ref):Object:this
Top method improvements (bytes):
         -36 (-2.22% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMemberContainerTypeSymbol:MergePartialMethods(Dictionary`2,DiagnosticBag)
         -30 (-2.91% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceUserDefinedOperatorSymbolBase:.ctor(int,String,SourceMemberContainerTypeSymbol,Location,SyntaxReference,SyntaxReference,SyntaxTokenList,DiagnosticBag,bool):this
         -30 (-14.93% of base) : System.Private.CoreLib.dasm - TaskFactory`1:GetDefaultScheduler(Task):TaskScheduler:this (3 methods)
         -28 (-6.54% of base) : System.Private.CoreLib.dasm - UnwrapPromise`1:.ctor(Task,bool):this (2 methods)
         -20 (-23.53% of base) : System.Private.CoreLib.dasm - Task:NotifyParentIfPotentiallyAttachedTask():this
         -18 (-1.16% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMemberContainerTypeSymbol:MakeTypeMembers(DiagnosticBag):Dictionary`2:this
         -17 (-0.40% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceNamedTypeSymbol:MakeOneDeclaredBases(ConsList`1,SingleTypeDeclaration,DiagnosticBag):Tuple`2:this
         -15 (-17.44% of base) : System.Private.CoreLib.dasm - Task:UpdateExceptionObservedStatus():this
         -13 (-6.99% of base) : System.Private.CoreLib.dasm - DateTime:ToFileTimeUtc():long:this
         -12 (-0.66% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceNamespaceSymbol:MakeNameToMembersMap(DiagnosticBag):Dictionary`2:this
         -12 (-0.62% of base) : Microsoft.CodeAnalysis.CSharp.dasm - MethodCompiler:BindMethodBody(MethodSymbol,TypeCompilationState,DiagnosticBag,byref):BoundBlock
         -12 (-1.34% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceDestructorSymbol:.ctor(SourceMemberContainerTypeSymbol,DestructorDeclarationSyntax,DiagnosticBag):this
         -12 (-3.36% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceConstructorSymbol:CheckModifiers(int,Location,DiagnosticBag):this
         -12 (-3.52% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMethodSymbol:PostDecodeWellKnownAttributes(ImmutableArray`1,ImmutableArray`1,DiagnosticBag,short,WellKnownAttributeData):this
         -12 (-1.22% of base) : System.DirectoryServices.dasm - ExceptionHelper:CreateSyncAllException(long,bool):SyncFromAllServersOperationException
         -12 (-3.38% of base) : System.Private.CoreLib.dasm - <>c:<OutputWaitEtwEvents>b__12_0(Action,Task):this
         -10 (-4.85% of base) : System.Private.CoreLib.dasm - Task:TaskConstructorCore(Delegate,Object,CancellationToken,int,int,TaskScheduler):this
         -10 (-16.13% of base) : System.Private.CoreLib.dasm - TaskFactory:GetDefaultScheduler(Task):TaskScheduler:this
         -10 (-19.23% of base) : System.Private.CoreLib.dasm - TaskScheduler:get_InternalCurrent():TaskScheduler
          -8 (-0.72% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - ConversionEasyOut:ClassifyPredefinedConversion(TypeSymbol,TypeSymbol):Nullable`1
Top method regressions (percentages):
           3 ( 0.06% of base) : System.Private.CoreLib.dasm - RuntimeType:InvokeMember(String,int,Binder,Object,ref,ref,CultureInfo,ref):Object:this
Top method improvements (percentages):
          -5 (-35.71% of base) : System.Private.CoreLib.dasm - Task:get_CreationOptions():int:this
          -5 (-27.78% of base) : System.Private.CoreLib.dasm - SystemThreadingTasks_FutureDebugView`1:get_CreationOptions():int:this
          -5 (-27.78% of base) : System.Private.CoreLib.dasm - SystemThreadingTasks_TaskDebugView:get_CreationOptions():int:this
         -20 (-23.53% of base) : System.Private.CoreLib.dasm - Task:NotifyParentIfPotentiallyAttachedTask():this
          -5 (-21.74% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMemberContainerTypeSymbol:get_IsStatic():bool:this
          -5 (-21.74% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMemberContainerTypeSymbol:get_IsSealed():bool:this
          -5 (-21.74% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMemberContainerTypeSymbol:get_IsAbstract():bool:this
          -5 (-21.74% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMethodSymbol:get_IsAbstract():bool:this
          -5 (-21.74% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMethodSymbol:get_IsStatic():bool:this
          -5 (-21.74% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMethodSymbol:get_IsSealed():bool:this
          -6 (-20.69% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - MethodLoadUnloadTraceData:get_IsDynamic():bool:this
          -6 (-20.69% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - MethodLoadUnloadTraceData:get_IsGeneric():bool:this
          -6 (-20.69% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - MethodLoadUnloadTraceData:get_IsJitted():bool:this
          -6 (-20.69% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - MethodLoadUnloadVerboseTraceData:get_IsDynamic():bool:this
          -6 (-20.69% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - MethodLoadUnloadVerboseTraceData:get_IsGeneric():bool:this
          -6 (-20.69% of base) : Microsoft.Diagnostics.Tracing.TraceEvent.dasm - MethodLoadUnloadVerboseTraceData:get_IsJitted():bool:this
          -5 (-19.23% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMemberContainerTypeSymbol:get_IsPartial():bool:this
          -5 (-19.23% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMethodSymbol:get_IsExtern():bool:this
          -5 (-19.23% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMethodSymbol:get_IsOverride():bool:this
          -5 (-19.23% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMethodSymbol:get_IsAsync():bool:this
110 total methods with Code Size differences (109 improved, 1 regressed), 196159 unchanged.
Completed analysis in 17.69s
```
Regression: https://www.diffchecker.com/pgRc3fEO

### CODE REVIEW

File: src/coreclr/src/jit/morph.cpp [hunk 1]
```diff
@@ -11611,6 +11611,68 @@ GenTree* Compiler::fgMorphFieldAssignToSIMDIntrinsicSet(GenTree* tree)
...
+// Arguments:
+//       tree - node to fold
+//
+// return value:
+//       A folded GenTree* instance or nullptr if something prevents folding.
+//
+
+GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+{
+    assert(tree->OperIs(GT_ADD, GT_MUL, GT_OR, GT_AND, GT_XOR));
```
Comment (tannergooding) [inline src/coreclr/src/jit/morph.cpp:11628]: Assert that the type isn't floating-point? Should there likewise be an assert that this isn't a checked operation?

Comment (EgorBo) [reply src/coreclr/src/jit/morph.cpp:11628]: > Should there likewise be an assert that this isn't a checked operation?

as far as I understand it's guarded via `gtOverflow`.

**Comment by AndyAyersMS** [comment]
Does this handle the multiple cases? If not, what might we do to try and address those?

```c##
x * 2 * 2 * 2;
x * 2 * y * 2;
```
Also the sub/div cases are potentially interesting, but probably outside the scope of this change:
```c#
x - 2 - 2;
x / 4 / 2;
```
As are mixed add/sub, mul/div, etc:
```
x - 2 + 2;
x * 2 / 2;
```

**Comment by EgorBo** [comment]
@AndyAyersMS 

> x * 2 * 2 * 2;

`X * C1 * C2 * C3` is handled but not when any C is `2` - it's morphed to `X << 1` too early. I'd honestly drop this behavior (it's going to be optimized to `lea` anyway. not sure about ARM tho).

> x * 2 * y * 2;

Yeah but I'd like to handle `(X op C1) op (Y op C2)` pattern separately if you don't mind 🙂 (we already support it for `GT_ADD`)

> x - 2 - 2;

it's handled because `X - C` is morphed into `X + (-C)`

> x * 2 / 2

this one is dangerous, e.g. `x * 3 / 3` for `x=715827883` produces `-715827882` (currently, unchecked context)

File: src/coreclr/src/jit/morph.cpp [hunk 2]
```diff
@@ -11611,6 +11611,68 @@ GenTree* Compiler::fgMorphFieldAssignToSIMDIntrinsicSet(GenTree* tree)
...
+//       A folded GenTree* instance or nullptr if something prevents folding.
+//
+
+GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+{
+    assert(varTypeIsIntegralOrI(tree->TypeGet()));
+    assert(tree->OperIs(GT_ADD, GT_MUL, GT_OR, GT_AND, GT_XOR));
+
+    // op1 can be GT_COMMA, in this case we're going to fold
+    // "COMMA(..., (X op C1)) op C2" to "COMMA(..., X op (C1 op C2))"
```
Comment (AndyAyersMS) [inline src/coreclr/src/jit/morph.cpp:11631]: Is this comment stale? It doesn't look like you handle the COMMA case.

Not sure it's worth the trouble. Might be a something to try and measure.

File: src/coreclr/src/jit/morph.cpp [hunk 3]
```diff
@@ -11611,6 +11611,68 @@ GenTree* Compiler::fgMorphFieldAssignToSIMDIntrinsicSet(GenTree* tree)
...
+    GenTree*   op1  = tree->gtGetOp1()->gtEffectiveVal(true);
+    genTreeOps oper = tree->OperGet();
+
+    if (!op1->OperIs(oper) || !tree->gtGetOp2()->IsCnsIntOrI() || !op1->gtGetOp2()->IsCnsIntOrI() ||
```
Comment (AndyAyersMS) [inline src/coreclr/src/jit/morph.cpp:11635]: Do we call this before or after `fgMorphSmpOpOptional` has moved constants to the right? Seems like it happens before, so maybe we should call `fgMorphCommutative ` from `fgMorphSmpOpOptional` after it calls `fgMoveOpsLeft`?

(perhaps try an example like `2 + x + 2` ?)

**Review by AndyAyersMS — COMMENTED** [review]
I always find the ordering of things in morph tricky to reason about.

File: src/coreclr/src/jit/morph.cpp (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/src/jit/morph.cpp:11631]: hm it does handle COMMA for me, it has `gtEffectiveVal(true)` means if it's COMMA - take gtOp2 of it
and preserves it in the end, will post here an example

File: src/coreclr/src/jit/morph.cpp (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/src/jit/morph.cpp:11635]: Looks like that rotation happens before `fgMorphCommutative`:
image

File: src/coreclr/src/jit/morph.cpp (hunk 2 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/morph.cpp:11631]: Maybe it's just the comment is misleading, it should be (in prefix notation)
```
(OP (COMMA(... (OP x C1))) C2)
==>
(COMMA(... (OP x (OP C1 C2))))
```
```

Comment (EgorBo) [reply src/coreclr/src/jit/morph.cpp:11631]: will change the comment, here is the case:

image

Comment (AndyAyersMS) [reply src/coreclr/src/jit/morph.cpp:11631]: Good. Sorry for the confusion.

**Comment by AndyAyersMS** [comment]
@EgorBo are you planning on any other changes (like revising that comment)?

**Comment by EgorBo** [comment]
> @EgorBo are you planning on any other changes (like revising that comment)?

Fixed the comment, I'll try to handle (X op C1) op (Y op C2) in a separate PR so it's finished from my end if there is no additional feedback.

**Review by AndyAyersMS — APPROVED** [review]
Looks good.

**Pull request merged**